### DISCUSSION TITLE
Phase 41.3 — AdversarialTrainer: PGD-AT, TRADES, AWP Deep-Dive

### DISCUSSION THREAD

**Discussion by web3guru888** [web3guru888/asi-build#834] [Show and tell]
# Phase 41.3 — AdversarialTrainer: Implementation Deep-Dive

## The Adversarial Training Paradigm

Adversarial training (Madry et al., 2018) reformulates the standard training objective as a min-max optimization:

```
min_θ E_{(x,y)} [ max_{‖δ‖≤ε} L(f_θ(x + δ), y) ]
```

The inner maximization finds worst-case perturbations; the outer minimization trains the model to be robust against them. This is the most effective empirical defense.

## PGD Adversarial Training (PGD-AT)

The canonical implementation:
1. For each batch (x, y):
   - Generate adversarial examples via PGD: x_adv = PGD(model, x, y, ε)
   - Compute loss on adversarial examples: L = CE(f(x_adv), y)
   - Update model weights: θ ← θ - η·∇_θ L

**Key insight from Madry et al.:** PGD with random restarts is a "universal first-order adversary" — if a model is robust to PGD, it's likely robust to any first-order attack.

## TRADES: Principled Robustness-Accuracy Trade-off

Zhang et al. (2019) decomposed the robust error into:
- **Natural error**: misclassification on clean data
- **Boundary error**: vulnerability to adversarial perturbation

TRADES loss:
```
L_TRADES = CE(f(x), y) + β · KL(f(x) ‖ f(x_adv))
```

The β parameter directly controls the trade-off:
- β = 0: standard training (no robustness)
- β → ∞: maximize robustness at any cost to accuracy
- β = 6: recommended value for ε=8/255 on CIFAR-10

**Critical difference from PGD-AT:** TRADES generates adversarial examples by maximizing KL divergence with the clean prediction (not cross-entropy with the label), leading to better boundary estimation.

## AWP: Weight-Space Robustness

Wu et al. (2020) showed that adversarial training overfits — robust training accuracy far exceeds robust test accuracy. AWP addresses this by also perturbing model **weights**:

```
For each batch:
  1. Standard adversarial training step → compute loss
  2. Perturb weights in the worst-case direction: θ' = θ + γ · ∇_θ L / ‖∇_θ L‖
  3. Recompute adversarial loss with perturbed weights
  4. Update original weights based on combined gradient
  5. Restore original weights
```

This flattens the loss landscape in weight space, improving generalization.

## Curriculum Adversarial Training

Rather than using full ε from the start, we gradually increase perturbation budget:

```
ε(t) = ε_max · min(1, t / T_warmup)
```

Benefits:
- More stable early training
- Better convergence to strong adversarial optima
- Avoids catastrophic loss divergence

## Robustness Benchmarks

| Method | Clean Acc | Robust Acc (PGD-20) | Robust Acc (AutoAttack) |
|--------|-----------|---------------------|------------------------|
| Standard | 95.0% | 0.0% | 0.0% |
| PGD-AT | 84.5% | 52.1% | 47.3% |
| TRADES (β=6) | 84.9% | 54.8% | 50.2% |
| TRADES + AWP | 85.6% | 57.1% | 52.8% |

*CIFAR-10, WideResNet-34-10, ε=8/255 Linf*

---
*See issue #827 for full specification. Part of [Phase 41 — Adversarial Robustness & Security Intelligence](../discussions/824).*
